trl/vllm_utils.py

```diff
@@ -0,0 +1,256 @@
+
+            except Exception as e:
+                self.app.logger.error(f"Error loading model: {str(e)}")
+                return jsonify({"error": str(e)}), 400
+
+        @self.app.route("/generate", methods=["POST"])
+        def generate():
+            if self.llm is None:
+                return jsonify({"error": "No model loaded. Load a model using the /load endpoint."}), 400
+
+            try:
+                # Parse input JSON data
+                data = request.get_json()
+
+                # Get prompts
+                prompts = data["prompts"]
+
+                # Get sampling params (optional)
+                sampling_params = SamplingParams(**data["sampling_params"]) if "sampling_params" in data else None
+
+                # Determine return type (default to 'text')
+                return_type = data.get("return_type", "text")
+                if return_type not in {"text", "tokens"}:
+                    return jsonify({"error": f"Invalid return_type '{return_type}'. Must be 'text' or 'tokens'."}), 400
+
+                # Generate completions
+                if is_conversational({"prompt": prompts[0]}):
+                    outputs = self.llm.chat(prompts, sampling_params=sampling_params)
+                else:
+                    outputs = self.llm.generate(prompts, sampling_params=sampling_params)
+
+                # Process output based on return_type
+                if return_type == "text":
+                    completions = [out.text for completions in outputs for out in completions.outputs]
+                elif return_type == "tokens":
+                    completions = [out.token_ids for completions in outputs for out in completions.outputs]
+
+                return jsonify(completions)
+
+            except Exception as e:
+                return jsonify({"error": str(e)}), 400
+
+        @self.app.route("/load_weights", methods=["POST"])
+        def load_weights():
+            if self.llm is None:
+                return jsonify({"error": "No model loaded. Load a model using the /load endpoint."}), 400
+            try:
+                # Parse binary data (weights file sent as bytes)
+                weights_data = request.data
+                buffer = io.BytesIO(weights_data)
+
+                # Load the state_dict from the buffer
+                state_dict = torch.load(buffer, weights_only=True).items()
+
+                # Update the model's weights
+                llm_model = self.llm.llm_engine.model_executor.driver_worker.model_runner.model
+                llm_model.load_weights(state_dict)
+
+                return jsonify({"status": "success", "message": "Model weights loaded."})
+
+            except Exception as e:
+                return jsonify({"status": "error", "message": str(e)}), 400
+
+    def run(self):
+        """Run the Flask server."""
+        self.app.run(host=self.host, port=self.port)
+
+
+class VLLMClient:
```

Comment:
Did you have to roll your own server / client because the existing [one](https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html?ref=blog.mozilla.ai) in `vllm` doesn't support model loading etc?

Reply:
Exactly